pkg/env/var.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package env makes it possible to track use of environment variables within procress
+// in order to generate documentation for these uses.
+package env
+
+import (
+	"os"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"istio.io/istio/pkg/log"
+)
+
+// Var holds the name and description of an environment variable.
+type Var struct {
+	// The name of the environment variable.
+	Name string
+
+	// The optional default value of the environment variable.
+	DefaultValue string
+
+	// Description of the environment variable's purpose.
+	Description string
+
+	// Hide the existence of this variable when outputting usage information.
+	Hidden bool
+
+	// Mark this variable as deprecated when generating usage information.
+	Deprecated bool
+}
+
+var allVars = make(map[string]Var)
+var mutex sync.Mutex
+
+// Returns a description of this process' environment variables, sorted by name.
+func GetVarDescriptions() []Var {
+	mutex.Lock()
+	sorted := make([]Var, 0, len(allVars))
+	for _, v := range allVars {
+		sorted = append(sorted, v)
+	}
+	mutex.Unlock()
+
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i].Name < sorted[j].Name
+	})
+
+	return sorted
+}
+
+// GetComplexVar returns the value of an environment variable, or a default value if the environment variable isn't set.
+//
+// In addition to the value, this function returns a boolean value indicating whether the variable existed or not.
+func GetComplexVar(v Var) (string, bool) {
```

Comment:
why "complex"? Should it just be a member function of `Var` to disambiguate with the methods that don't operate on `Var`?